doc/run-cloud-build-locally.md

```diff
@@ -161,37 +158,221 @@ Execute the HAPI search test:
+  -e PARQUET_PATH=/workspace/e2e-tests/JDBC_HAPI \
+  -e OUTPUT_PARQUET_VIEW_PATH=/workspace/e2e-tests/JDBC_HAPI/VIEWS_TIMESTAMP_1 \
+  -e JDBC_FETCH_SIZE=1000 \
+  -v $(pwd):/workspace \
+  --network host \
+  ${_REPOSITORY}/batch-pipeline:${_TAG}
+```
+
+### 2. Run e2e tests for JDBC mode
+
+```bash
+./e2e-tests/pipeline_validation.sh e2e-tests/ JDBC_HAPI JDBC_HAPI_FHIR_JSON http://localhost:9002/fhir
+```
+
+## Bulk Export Mode
+
+**Note:** Only run this section if debugging Bulk Export mode; otherwise skip.
+
+### 1. Run batch pipeline for Bulk Export mode
+
+```bash
+docker run --rm \
+  -e FHIR_FETCH_MODE=BULK_EXPORT \
+  -e FHIR_SERVER_URL=http://localhost:9001/fhir \
+  -e PARQUET_PATH=/workspace/e2e-tests/BULK_EXPORT \
+  -e OUTPUT_PARQUET_VIEW_PATH=/workspace/e2e-tests/BULK_EXPORT/VIEWS_TIMESTAMP_1 \
+  -v $(pwd):/workspace \
+  --network host \
+  ${_REPOSITORY}/batch-pipeline:${_TAG}
+```
+
+### 2. Run e2e tests for Bulk Export mode
+
+```bash
+./e2e-tests/pipeline_validation.sh e2e-tests/ BULK_EXPORT BULK_EXPORT_FHIR_JSON
+```
+
+## Controller and Spark Mode
 
-### 7. Clean up containers
+**Note:** Only run this section if debugging Controller and Spark mode;
+otherwise skip.
 
-When finished, bring everything down to avoid port conflicts:
+### 1. Turn down sink servers
 
 ```bash
 docker compose -f docker/sink-compose.yml -p sink-server-search down -v
 docker compose -f docker/sink-compose.yml -p sink-server-jdbc down -v
-docker compose -f docker/hapi-compose.yml -p hapi-compose down -v
 ```
 
-Also stop the controller/spark Compose services and other environment-specific
-stacks as needed.
+### 2. Create views database
+
+The `views` database is used for creating flat views from ViewDefinitions.
+
+```bash
+docker run --rm --network host postgres psql -U admin -d postgres -h localhost -p 5432 -c 'CREATE DATABASE views;'
+```
+
+### 3. Build the e2e controller/spark image
+
+```bash
+docker build -t ${_REPOSITORY}/e2e-tests/controller-spark:${_TAG} -f e2e-tests/controller-spark/Dockerfile .
+```
+
+### 4. Bring up controller and Spark containers
+
+```bash
+PIPELINE_CONFIG=/workspace/docker/config DWH_ROOT=/workspace/e2e-tests/controller-spark/dwh docker compose -f docker/compose-controller-spark-sql-single.yaml up --force-recreate -d
+```
+
+### 5. Run e2e test for controller and Spark
+
+```bash
+docker run --rm -v $(pwd):/workspace ${_REPOSITORY}/e2e-tests/controller-spark:${_TAG}
+```
+
+### 6. Bring down controller and Spark containers
+
+```bash
+docker compose -f docker/compose-controller-spark-sql-single.yaml down -v
+```
+
+## FHIR to FHIR Sync Mode
+
+**Note:** Only run this section if debugging FHIR to FHIR sync mode; otherwise
+skip.
+
+### 1. Launch HAPI FHIR Sink Server for controller
+
+```bash
+SINK_SERVER_NAME=sink-server-controller SINK_SERVER_PORT=9001 docker compose -f docker/sink-compose.yml -p sink-server up --force-recreate -d
+```
+
+### 2. Bring up pipeline controller for FHIR to FHIR sync
```

Comment:
Please add a tip that people can use the controller from the previous step if they have it running (instead of bringing down/up).